leetcode/271.EncodeDecodeStrings.cpp

```diff
@@ -0,0 +1,53 @@
+class Codec
+{
+public:
+    // 271. Encode and Decode Strings: https://leetcode.com/problems/encode-and-decode-strings/
+    // Encode: Space Complexity: O(n*l); Time Complexity: O(n*l)
+    // Decode: Space Complexity: O(n*l); Time Complexity: O(n*l)
+
+    // Encodes a list of strings to a single string.
+    string encode(const vector<string> &strs)
+    {
+        string result;
+        // Time complexity: number of words
+        // Space complexity: O(n*l)
+        for (int i = 0; i < strs.size(); i++)
+            // for each word, store a combination of the
+            // number of characters, encode string, and the word itself
+            // this makes it easy to decode later
+            result += to_string(strs[i].size()) + mEncodeStr + strs[i];
+
+        return result;
+    }
+
+    // Decodes a single string to a list of strings.
+    vector<string> decode(const string &s)
+    {
+        vector<string> result;
+        string word;
+        int i = 0;
+        // Time complexity: O(n*l) number of words that were decoded and length of each word
+        // Space complexity: O(n) for the result vector
+        while (i < s.size())
+        {
+            string tmp;
+            while (s[i] != mEncodeStr[0] && i < s.size())
+            {
+                // extract the digits prior to the encoding string
+                tmp += s[i];
+                i++;
+            }
+
+            // use the length of the word to identify indices of substring for word
+            int lengthOfWord = stoi(tmp); // O(k) -- number of digits of word
+
+            result.push_back(s.substr(i + mEncodeStr.size(), lengthOfWord));  // O(l) length of word
+            i += mEncodeStr.size() + lengthOfWord;
+        }
+
+        return result;
+    }
+
+private:
+    string mEncodeStr{"@@"};
```

Comment:
this seems like it is a constant. If so, you can use the following at the top of the file.

```
inline constexpr string_view kEncodeStr = "@@"
```

Reply:
Also - why does this need to be `@@`? Can we just do `@`?

Reply:
You're right. We could do just `@` .

Originally, I had `@@` because I wasn't using the length of the `word` to encode the word. So, I had to make sure `mEncodeStr` was long enough to avoid clashing with possible text combinations of the actual words.

However, now that we can use the length of the `word` to decode the string, clashing with text combinations won't happen. So, length of `mEncodeStr` doesn't matter anymore.

Reply:
@mchen17, why `kEncodeStr` instead of `mEncodeStr`? Is that a style best practice that I'm not aware of? Or a typo?

Reply:
Verbally discussed and I also found this style guide: https://google.github.io/styleguide/cppguide.html#Constant_Names